File: src/Aiko.Core/PersonaPluginOutput.cs
namespace Aiko.Core;

/// Where the bridge puts the persona plugin it builds, and which old copies it clears (D-201).
///
/// Claude Code runs "Aiko.Bridge.exe plugin aiko-persona", copies the folder the bridge prints into
/// its own cache and never reads the folder again. Each content gets its own folder named after
/// its hash, so a changed persona never half-overwrites a folder Claude Code may be copying.
public static class PersonaPluginOutput
{
    public const string Verb = "plugin";

    /// How long an old folder stays. Another session of the same environment may be copying it
    /// right now; an hour is far longer than any copy takes.
    public static readonly TimeSpan KeepOldFor = TimeSpan.FromHours(1);

    /// One folder per environment, named like its limit snapshot, so the two never disagree.
    public static string EnvironmentFolder(AikoFolders folders, string configDirectory) =>
        Path.Combine(folders.PersonaPluginsFolder, SnapshotName.For(configDirectory));

    public static string VersionFolder(AikoFolders folders, string configDirectory, string contentHash) =>
        Path.Combine(EnvironmentFolder(folders, configDirectory), contentHash);

    /// Whether the command asks for a plugin this bridge can build.
    public static bool IsRequest(IReadOnlyList<string> args) =>
        args is [Verb, PersonaPlugin.Name];

    /// Old folders to delete: every other version older than KeepOldFor. Folders that are not a
    /// version at all, such as a temporary folder of a build still running, are left alone.
    public static IReadOnlyList<string> StaleFolders(
        IEnumerable<(string Name, DateTimeOffset Written)> folders, string currentHash, DateTimeOffset now) =>
        folders
            .Where(f => f.Name != currentHash && IsHash(f.Name) && now - f.Written > KeepOldFor)
            .Select(f => f.Name)
            .ToList();

    private static bool IsHash(string name) =>
        name.Length == 12 && name.All(c => c is >= '0' and <= '9' or >= 'a' and <= 'f');
}

File: mac/Sources/AikoKit/PersonaPluginOutput.swift
import Foundation

/// Where the bridge puts the persona plugin it builds, and which old copies it clears (D-201).
///
/// Claude Code runs the bridge with "plugin aiko-persona", copies the folder the bridge prints into
/// its own cache and never reads the folder again. Each content gets its own folder named after
/// its hash, so a changed persona never half-overwrites a folder Claude Code may be copying.
public enum PersonaPluginOutput {
    public static let verb = "plugin"

    /// How long an old folder stays. Another session of the same environment may be copying it
    /// right now; an hour is far longer than any copy takes.
    public static let keepOldFor: TimeInterval = 3600

    /// One folder per environment, named like its limit snapshot, so the two never disagree.
    public static func environmentFolder(_ folders: AikoFolders, _ configDirectory: String) -> String {
        folders.platform.join(folders.personaPluginsFolder, SnapshotName.forConfigDirectory(configDirectory))
    }

    public static func versionFolder(
        _ folders: AikoFolders, _ configDirectory: String, _ contentHash: String
    ) -> String {
        folders.platform.join(environmentFolder(folders, configDirectory), contentHash)
    }

    /// Whether the command asks for a plugin this bridge can build.
    public static func isRequest(_ args: [String]) -> Bool {
        args == [verb, PersonaPlugin.name]
    }

    /// Old folders to delete: every other version older than keepOldFor. Folders that are not a
    /// version at all, such as a temporary folder of a build still running, are left alone.
    public static func staleFolders(
        _ folders: [(name: String, written: Date)], currentHash: String, now: Date
    ) -> [String] {
        folders
            .filter { $0.name != currentHash && isHash($0.name) && now.timeIntervalSince($0.written) > keepOldFor }
            .map(\.name)
    }

    private static func isHash(_ name: String) -> Bool {
        name.count == 12 && name.allSatisfy { $0.isNumber && $0.isASCII || ("a"..."f").contains($0) }
    }
}
